itou/common_apps/organizations/models.py

```diff
@@ -74,6 +78,18 @@ def has_member(self, user):
     def has_admin(self, user):
         return self.active_admin_members.filter(pk=user.pk).exists()
 
+    def expire_invitations(self, user):
+        expired = self.invitations.pending().filter(Q(email=user.email) | Q(sender=user)).update(validity_days=0)
+        logger.info(
+            "Expired %(expired)d invitations to %(org_model)s %(org_id)d for user_id=%(user_id)d.",
```

Comment:
J'aurais bien vu des appels différents (et des logs différents) pour les deux différents cas `expire_invitations_to_user`/`expire_invitations_from_user`.

Reply:
(soit 2 méthodes, soit `expire_invitations(*, from_user=None, to_user=None)` avec `assert (from_user is None) != (to_user is None)` (ou équivalent)

Reply:
Pour quelle raison ?
Lorsqu’on retire un utilisateur d’une structure, on veut bien retirer les invitations pour cet utilisateur et les invitations qu’il a émises. Lors de l’ajout ou de l’activation, je m’attends à ce qu’il n’y ait aucune invitation émise par l’utilisateur, donc les expirer ne me semble pas poser de problème.
Ce serait pour les logs ?

Reply:
Oui, c'est principalement pour les logs.
Mais si on rajoute un if pour les logs on pourrait également le mettre pour la requête SQL.
Et enfin pour la compréhension je trouve cela plus clair pour comprendre ce que le code fait.
`expire_invitations(from_user=user)` permet de savoir tout de suite qu'on désactive les invitations envoyées.

Reply:
C’est qu’on va devoir appeler de pair `expire_invitations(from_user=user)` et `expire_invitations(to_user=user)` au moins au moment du retrait d’un membre, et probablement à l’ajout pour la compatibilité avec l’ancien code qui n’expirait pas les invitations envoyées par l’utilisateur à la désactivation.

Reply:
Autrement dit, toujours appeler les deux méthodes.
Si on veut des logs différents (ce qui me semble légitime), je pousserais plutôt pour une seule méthode qui fait deux logs et deux requêtes ?

Reply:
Effectivement à la désactivation on appellera les deux. Donc finalement une seule méthode avec les deux arguments me semble bien.
A l'acceptation, on appellerait: `expire_invitations(to_user=user)` pour invalider les éventuelles autres invitations et à la désactivation, on appelle `expire_invitations(to_user=user, from_user=user)` ou `expire_invitations(to_user=user, include_sent_invitations=True)`.

Reply:
Pour moi, à l’acceptation, on veut également supprimer les invitations envoyées (au moins pendant 14 jours).

Reply:
On pourrait mais a priori si l'utilisateur vient d'accepter l'invitation il n'a pas pu envoyer d'invitations ? :thinking: 
Après ça ne fait qu'une petite requête SQL et un log indiquant qu'on a rien désactivé avec le bénéfice d'enlever une option à la méthode donc ce n'est pas dramatique mais je trouve cela un peu bizarre :man_shrugging:

Reply:
Une autre solution serait effectivement une migration pour les attraper et les corriger.

Reply:
J’ai remis la solution ceinture bretelle, normalement on devrait être tranquilles sans les pinces à linge.